### ISSUE TITLE
from_hdf5 function that uses filenames rather than files

### ISSUE THREAD

**Issue by mrocklin** [blaze/dask#922]
The current approach to load data from HDF5 is to load in an h5py or netcdf4 Dataset object and pass that to `from_array`.  This is efficient, but also includes the file pointer within the dask graph.  This fails if you need to serialize the graph, such as is necessary in distributed computing.

It might be wise to instead create an explicit `da.from_hdf5` function that only stored the filename, datapath, and slice information within the graph.  

It would also be nice if this API was exposed up to `xarray` (cc @shoyer).

cc @rabernat @pwolfram

**Comment by jcrist** [comment]
A few ideas:

- Expanding your idea above, signature could be `da.from_hdf5(filepath, datapath)`. Slice is composed later, as it feels more natural to index (also, `:` only works in getitem). Example:
```python
z1 = da.from_hdf5('foo.h5', '/x/y/z1')[:, :100]
z2 = da.from_hdf5('foo.h5', '/x/y/z2')[:, :100]
```

- Another approach would be to make a "lazy" version of the h5py File objects that don't open the file pointer until explicitly asked to (or closes and reopens the file pointer when serialized). Example:
```python
x = da.from_hdf5('foo.h5')
z1 = x['/x/y/z1'][:, :100]
z2 = x['/x/y/z2'][:, :100]
```

**Comment by mrocklin** [comment]
I agree that the slicing should occur later, but anything that gathers data from an array-thing will have to include slices within the generated tasks.  For example `from_array` includes many tasks, each of which slice from the underlying array.

The lazy/serializable h5py File object solution sounds like a nice one.

**Comment by shoyer** [comment]
@rabernat did some related experiments to avoid system limits on the number of open files. Unfortunately, getting satisfactory performance was a challenge: https://github.com/pydata/xarray/pull/524

It seems critical to avoid repetitively opening and closing files in every single task. One solution I imagined but never got around to implementing would be to create a single "open file" task, followed by a bunch of dependent "get data" tasks and a "close file" task that depends on the data. However, it's not clear how you would actually ensure that "close file" gets called (at least with dask), given that none of the computation depends on it.

As for getting this into xarray, the main challenge is that we use external netCDF libraries instead of reading them directly. But if you can figure this out for h5py, then it should be quite easy to hook into [h5netcdf](https://github.com/shoyer/h5netcdf). I don't think we would even need to add a dask dependency to add a `lazy` argument to `h5netcdf.File`.

**Comment by mrocklin** [comment]
Perhaps we cache the opening per worker process and just don't close things?  Perhaps we close things after waiting for 1s in a separate thread.

**Comment by mrocklin** [comment]
Here is a solution with a lazy h5py dataset. 

https://gist.github.com/mrocklin/1f61ec7e669563f05505

```python
In [1]: from hdf5lazy import Dataset

In [2]: %timeit dset = Dataset('foo.hdf5', 'x')
The slowest run took 52.19 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 10.4 µs per loop
```

**Comment by mrocklin** [comment]
@rabernat @shoyer any thoughts on this approach?  (the code linked to above is, I claim, very approachable.)

@pwolfram noted that the interface for netcdf3 is a bit different.  @shoyer can you think of a way to generalize the code above?  Or should we just write a few of these for each for the different libraries?

**Comment by rabernat** [comment]
This seems similar to, but far cleaner than, to the [hack job](https://github.com/pydata/xarray/pull/524/files#diff-fbb53316626aea75814a3e110a721988R33) I tried on the xray ScipyNetcdfWrapper. My goal there was to get around the system ulimit on open files, which was limiting my ability to open large mfdatasets.

Your docstring says that the class limits the number of open files via the cache. But this seems to apply only when the same file would otherwise be opened multiple times, e.g. maybe when chunking. My approach went further and actually attempted to close the file after each read and reopen once data was requested. That is what gave the big performance penalty.

It seems like your code solves two important problems:
- reducing the number of open files when chunking a single large dataset
- allowing file object to be serialized for distributed

It would be pretty straightforward to extend this approach to the netcdf engines in xarray, although @shoyer is much more qualified to judge that.

**Comment by mrocklin** [comment]
Notably my solution doesn't close files explicitly. I'm actually mostly ok with this.  We could make it more thorough by adding in a timed callback if people feel concerned about having the dangling file handle.

**Comment by shoyer** [comment]
Serializing the file object does seem pretty straightforward -- you simply need custom methods to overload how pickle works. We could certainly put that into xray (on generic data stores), and maybe even push it into upstream libraries.

This doesn't solve the "too many files" annoyance, though.

**Comment by mrocklin** [comment]
Meaning you sometimes have thousands of hdf5 files?

Perhaps we add an LRU mechanism to the file cache and open files on demand.

**Comment by shoyer** [comment]
Yes, it's not uncommon to encounter one file per time step.

**Comment by mrocklin** [comment]
I uploaded the hdf5lazy hack to PyPI in case anyone wants to try it out.  I added a wrapper for NetCDF4.  It doesn't handle the many file case yet.  It's also very shoddy work.

https://github.com/mrocklin/hdf5lazy

    $ pip install hdf5lazy

```python
In [1]: from glob import glob
In [2]: filenames = ['/home/mrocklin/data/ecmwf/' + fn for fn in sorted(glob('*.nc3'))] 

In [3]: from hdf5lazy import NetCDFDataset
In [4]: dsets = [NetCDFDataset(fn, 't2m') for fn in filenames]

In [5]: import dask.array as da
In [6]: arrays = [da.from_array(dset, chunks=dset.shape) for dset in dsets]
In [7]: stack = da.concatenate(arrays, axis=0)

In [8]: stack[0].sum().compute()
Out[8]: 287901475.39664292

In [9]: from distributed import Executor
In [10]: e = Executor('192.168.1.141:8786')

In [11]: stack[0].sum().compute(get=e.get)
Out[11]: 287901475.39664292

In [12]: stack.sum().compute(get=e.get)
Out[12]: 423832855970.39758

In [13]: stack
Out[13]: dask.array<concate..., shape=(1464, 721, 1440), dtype=int16, chunksize=(4, 721, 1440)>
```

**Comment by pwolfram** [comment]
There appears to be a balance between opening/closing files on demand vs having too many open files at a time via cache.  One solution @mrocklin proposed, if I understand correctly, would be to limit the size of the cache based on closing the last used files.  I'm assuming this is one of the first problems that needs solved before xarray can be connected to distributed, right?

**Comment by mrocklin** [comment]
Yes, so the problem as I see it is that, for serialization and open-file reasons we want to use a function like the following:

```python
def get_chunk_of_array(filename, datapath, slice):
    with netCDF4.Dataset(filename) as f:
        return f.variables[datapath][slice]
```

However, this opens and closes many files, which while robust, is slow.  We can alleviate this by maintaining an LRU cache in a global variable so that it is created separately per process.

```python
from toolz import memoize

cache = LRUDict(size=100, on_eviction=lambda file: file.close())

netCDF4_Dataset = memoize(netCDF4.Dataset, cache=cache)

def def get_chunk_of_array(filename, datapath, slice):
    f = netCDF4_Dataset(filename)
    return f.variables[datapath][slice]
```

I'm happy to supply the `memoize` function with `toolz` and an appropriate `LRUDict` object with other microprojects that I can publish if necessary.

We would then need to use such a function within the dask.array and xarary codebases.

Anyway, that's one approach.  Thoughts welcome.

**Comment by shoyer** [comment]
Yes, using such an LRU cache seems like a sane solution to me. It will
require some refactoring of xarray's backend system to make this work,
though -- we currently only call netCDF4.Dataset once, when we open files.

**Comment by jakirkham** [comment]
I'm trying to understand how PR ( https://github.com/dask/distributed/pull/606 ) relates to this if at all. Could someone please help me understand?

**Comment by jakirkham** [comment]
Maybe the answer is this ( https://github.com/dask/distributed/pull/620 )?

**Comment by mrocklin** [comment]
@jakirkham yes.  Custom serialization will allow us to special case certain datatypes like `h5py.Dataset`s, allowing them to live in the dask graph and also work with the distributed scheduler.  There are still other problems that need to be solved to make HDF5 and dask.distributed work cleanly together, like having a shared file system, but this was the main blocker.

**Comment by jakirkham** [comment]
> Custom serialization will allow us to special case certain datatypes like `h5py.Datasets`, allowing them to live in the dask graph and also work with the distributed scheduler.

That sounds great! Have been wanting to use dask more generally for a while, but this has been a bit of a blocker. If there is anything I can do to help, please let me know.

> There are still other problems that need to be solved to make HDF5 and dask.distributed work cleanly together, like having a shared file system, but this was the main blocker.

So there are some cases where this should be fine like working on single machine or using NFS. Though implementing a generic shared file system to handle a wide range for storage use cases feels a bit ill-defined IMHO. Could you please elaborate a bit more what use cases you are try to target with this?

**Comment by mrocklin** [comment]
When I said shared file system I meant something like an NFS that presents a POSIX compatible file system that all machines can view equally.

**Comment by Exteris** [comment]
@mrocklin I took your hdf5lazy approach a bit further by wrapping all h5py objects to support `__{set,get}state__` and to remember how we opened the file (especially mode is important).
An LRU cache is also implemented. This allows me to replace `import h5py` with `import h5pickle as h5py` and have working serialization in the dask distributed scheduler.

See the code at https://github.com/Exteris/h5pickle

This works very well for me in dask. Would you mind taking a look at the code? Perhaps this could be useful in distributed.

ps. Thanks for making dask so awesome

**Comment by shoyer** [comment]
@Exteris Nice, that looks pretty handy! It might be worth a call-out in the dask docs somewhere.

**Comment by vigji** [comment]
I write here as I encountered the problem of too many .h5 files opened when creating a task array in the following way:
```
dsets = [h5py.File(f, mode='r')['/stack_4D'] for f in
                             sorted(list(f.glob("*.h5")))]
arrays = [da.from_array(dset) for dset in dsets]
data = da.concatenate(arrays, axis=0)
```

Is there now a solution for this problem? Or any hack someone can recommend?

**Comment by Exteris** [comment]
@vigji can you provide us with the error message? It might be as simple as increasing the number of open file descriptors you can have. You can see the current limits with
`ulimit -Hn` and `ulimit -Sn` for hard and soft limits. If one of these numbers is close to the number of files you have you could try increasing it. See https://www.tecmint.com/increase-set-open-file-limits-in-linux/

**Comment by vigji** [comment]
Thanks, now I managed to make it working by changing the OS limits. Still wondering whether there's a cleaner option!

**Comment by Exteris** [comment]
There is no way to open more files than the OS allows (but the default limit can be too low for some applications as you see now). It is not so bad to change this as necessary.

Instead of concatenating the files manually you could also use the dask function for this: https://docs.dask.org/en/latest/dataframe-api.html#dask.dataframe.read_hdf

**Comment by vigji** [comment]
Thanks a lot!
